README.md
# Scientific Python Interactive Data Acoustic Modeling

This python project finds the average RT60 time for the sound of a hand clap. The ideal difference for speech intelligibility is 0.5 seconds, any difference value over 0.5 seconds means vocal intelligibility is reduced in the location the audio was recorded in.

After all the requirements have been installed, run the controller.py file to use the program. Navigate through each graph by closing the previous one. Pressing the extra data button will show you a graph with more visual information about the sound.


CHANGELOG.md
### December 8, 2023

The audio file was added along with the function to calculate the length of the audio file.

### December 9, 2023

MVC format was implemented and tested, .toml file was added, and the majority of calculating the RT60 value of the audio file was added, and the function to display the waveform of the audio file.

### December 10, 2023

Finishing touches were added to the MVC layout. This included the button for extra data, the RT60 difference value, the combined RT60 graphs, and the CHANGELOG.md file was added.


view.py
from model import *
import os
from tkinter import filedialog
import tkinter as tk
import numpy as np
from PIL import Image, ImageTk
import wave
import matplotlib.pyplot as plt


def load_audio_file():
    # filedialogue allows user to pick file w button
    # askopenfilename is just the request
    file_types = ["*.wav", "*.mp3"]
    file_path = filedialog.askopenfilename(title="Select Audio File", filetypes=[("Audio Files", file_types)])
    #checks if user selects a file
    if file_path:
        debugg(f"file_path: {file_path}")
        global wave_file
        wave_file = clean_file(file_path)

        extra_data_button()
        plot_rt60(wave_file)
        show_audio_length(wave_file)
        show_rt60(wave_file)
        plot_wave(wave_file)
        plot_all_rt60s()


def clean_file(file_path):
    # audio_file = open(file_path, 'rb')
    # debugg(f"audio_file: {audio_file.name}")
    # Checking if file is .mp3 and converting to .wav if so
    if (os.path.splitext(file_path))[1] == ".mp3":
        debugg(f"File extension: {(os.path.splitext(file_path))[1]}")
        wave_file = mp3_to_wav(file_path)
    else:
        # wave_file = wave.open(file_path, 'rb')
        wave_file = file_path
    debugg(f"wave_file: {wave_file}")
    # Stripping channels and metadata
    # wave_file = audio_stripper(wave_file)
    return wave_file

def plot_rt60(file_path):
    # Specify a list of frequencies
    frequencies = [250, 1000, 5000]

    # Create a color map for different frequencies
    color_map = {250: "blue", 1000: "green", 5000: "red"}

    # Calculate and plot RT60 for each frequency
    for frequency in frequencies:
        calculate_rt60(file_path, frequency, color_map.get(frequency))

def plot_all_rt60s():
    # Specify a list of frequencies
    frequencies = [250, 1000, 5000]
    colors = {250: "blue", 1000: "green", 5000: "red"}

    combine_rt60s(frequencies, colors)

def plot_frequency_spectrum(file_path):
    # Load audio file
    wave_file = wave.open(file_path, "rb")
    framerate = wave_file.getframerate()
    frames = wave_file.readframes(-1)
    signal = np.frombuffer(frames, dtype=np.int16)

   # Create a Matplotlib figure
    fig, ax = plt.subplots(figsize=(8, 6))

    # Calculate the Short-Time Fourier Transform (STFT)
    result = plt.specgram(signal, NFFT=1024, Fs=framerate, noverlap=512, cmap="viridis")
    Sxx = result[2]

    # Displays for the chart
    plt.xlabel("Time (s)")
    plt.ylabel("Frequency (Hz)")
    plt.title("Clap Audio")

    plt.show()

def plot_wave(file_path):
    # Load audio file
    wave_file = wave.open(file_path, "rb")
    framerate = wave_file.getframerate()
    frames = wave_file.readframes(-1)
    signal = np.frombuffer(frames, dtype=np.int16)

    # Create a Matplotlib figure
    fig, ax = plt.subplots(figsize=(8, 6))

    # plot wave
    time = np.linspace(0, len(signal) / framerate, num=len(signal))
    plt.plot(time, signal, color="red")

    # Displays for the chart
    plt.xlabel("Time (s)")
    plt.ylabel("Amplitude")
    plt.title("Clap Audio")

    plt.show()

def load_extra_data():
    plot_frequency_spectrum(wave_file)

def extra_data_button():
    # Create a button to load extra data
    extra_data_button = tk.Button(root, text="Extra Data", command=load_extra_data)
    extra_data_button.pack(pady=20, padx=30)

def file_gui():
    # Setting up main window with tkinter
    global root
    root = tk.Tk()
    root.title("Waveform")

    # Create a button to load an audio file
    load_button = tk.Button(root, text="Load Audio File", command=load_audio_file)
    load_button.pack(pady=20, padx=30)

    # Create a label to display the spectrogram image
    global label
    label = tk.Label(root)
    label.pack()

    # call 
    root.mainloop()

def show_rt60(wave_file):
    # Showing rt60 value reduce to 0.5 seconds
    rt60_text = tk.Text(root)
    rt60_text.insert(tk.END, f"Difference: {round(abs(rt60_difference(wave_file)), 2)}")
    rt60_text.pack()

def show_audio_length(wave_file):
    # Showing audio file length
    audio_length_text = tk.Text(root)
    audio_length_text.insert(tk.END, f"Audio file length: {round(get_audio_length(wave_file), 2)} s")
    audio_length_text.pack()


wav_handler.py
import os
from pydub import AudioSegment
from tkinter import filedialog
import tkinter as tk
import numpy as np
from PIL import Image, ImageTk
import wave
import matplotlib.pyplot as plt

class AudioAnalyzer:
    def __init__(self):
        self.file_path = None
        self.signal = None

    def mp3_to_wav(self, mp3_audio):
        dst, ext = os.path.splitext(mp3_audio)
        wav_audio = dst + '.wav'

        sound = AudioSegment.from_mp3(mp3_audio)
        sound.export(wav_audio, format='wav')
        return sound

    def get_audio_files(self):
        audio_files = []
        for file in os.listdir():
            if file.endswith('.wav'):
                audio_files.append(file)
            elif file.endswith('.mp3'):
                self.mp3_to_wav(file)
                audio_files.append(file)
        return audio_files

    def load_audio_file(self):
        file_types = ["*.wav", "*.mp3"]
        self.file_path = filedialog.askopenfilename(title="Select Audio File", filetypes=[("Audio Files", file_types)])

        if self.file_path:
            self.signal = self.plot_wave(self.file_path)
            show_freq_button.pack()

    def plot_wave(self, file_path):
        wave_file = wave.open(file_path, 'rb')
        framerate = wave_file.getframerate()
        frames = wave_file.readframes(-1)
        signal = np.frombuffer(frames, dtype=np.int16)

        fig, ax = plt.subplots(figsize=(8, 6))

        time = np.linspace(0, len(signal) / framerate, num=len(signal))
        plt.plot(time, signal, color='red')

        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
        plt.title('Clap Audio')

        image_path = 'waveform_plot.png'
        plt.savefig(image_path)

        img = Image.open(image_path)
        photo = ImageTk.PhotoImage(img)
        label.config(image=photo)
        label.image = photo

        # Return the signal array
        return signal

    def plot_frequency_spectrum(self):
        if self.file_path:
            wave_file = wave.open(self.file_path, 'rb')
            framerate = wave_file.getframerate()
            frames = wave_file.readframes(-1)
            signal = np.frombuffer(frames, dtype=np.int16)

            fig, ax = plt.subplots(figsize=(8, 6))

            result = plt.specgram(signal, NFFT=1024, Fs=framerate, noverlap=512, cmap='viridis')
            Sxx = result[2]

            plt.xlabel('Time (s)')
            plt.ylabel('Frequency (Hz)')
            plt.title('Clap Audio')

           
            plt.show()

# Setting up the main window with tkinter
root = tk.Tk()
root.title("Waveform")

audio_analyzer = AudioAnalyzer()

# button to load an audio file
load_button = tk.Button(root, text="Load Audio File", command=audio_analyzer.load_audio_file)
load_button.pack(pady=20, padx=30)

#label to display the waveform image
label = tk.Label(root)
label.pack()

#  button to show frequency graph
show_freq_button = tk.Button(root, text="Show Frequency", command=audio_analyzer.plot_frequency_spectrum)

# Call 
root.mainloop()


model.py
# This file keeps the state information of the system

# from view import *
import os
import numpy as np
import matplotlib.pyplot as plt
from pydub import AudioSegment
from scipy.io import wavfile


# Debug printing function
def debugg(fstring):
    # print(fstring)
    pass

# Converting .mp3 file to .wav
def mp3_to_wav(mp3_audio):
    dst, ext = os.path.splitext(mp3_audio)
    wav_audio = dst + ".wav"
    
    sound = AudioSegment.from_mp3(mp3_audio)
    sound.export(wav_audio, format="wav")
    debugg(f"sound: {wav_audio}")
    return wav_audio

# Returns all .wav and .mp3 files as .wav files
def get_audio_files():
    # List of audio files
    audio_files = []
    # Getting audio files in audio directory
    for file in os.listdir():
        if file.endswith(".wav"):
            audio_files.append(file)
        elif file.endswith(".mp3"):
            mp3_to_wav(file)
            audio_files.append(file)
    return audio_files

# Strips extra channels and metadata
def audio_stripper(audio_file):
    raw_audio = AudioSegment.from_file(audio_file, format="wav")
    mono_wav = raw_audio.set_channels(1)
    mono_wav.export(audio_file, format="wav")
    debugg(f"mono_wav path: {audio_file}")
    return audio_file

# Returns length of audio file
def get_audio_length(audio_file):
    samplerate, data = wavfile.read(audio_file)
    debugg(f"number of channels = {data.shape[len(data.shape) - 1]}")
    debugg(f"sample rate = {samplerate}Hz")
    length = data.shape[0] / samplerate
    return length

# Find a nearest value
def find_nearest_value(array, value):
    array = np.asarray(array)
    debugg(f"array {array[:10]}")
    idx = (np.abs(array - value)).argmin()
    debugg(f"idx {idx}")
    debugg(f"array[idx] {round(array[idx], 2)}")
    return array[idx]

def find_target_frequency(freqs, ideal_freq):
    for x in freqs:
        if x > ideal_freq:
            break
    return x

def frequency_check(frequency):
    debugg(f"freqs {freqs[:10]}")
    target_frequency = find_target_frequency(freqs, frequency)
    debugg(f"target_frequency {round(target_frequency, 2)}")
    index_of_frequency = np.where(freqs == target_frequency) [0][0]
    debugg(f"index_of_frequency {index_of_frequency}")
    
    # Find a sound data for a particular frequency
    data_for_frequency = spectrum[index_of_frequency]
    debugg(f"data_for_frequency {data_for_frequency[:10]}")
    
    # Change a digital signal for a values in decibles
    data_in_db_fun = 10 * np.log10(data_for_frequency)
    return data_in_db_fun

def calculate_rt60(audio_file, frequency, color):
    # Getting sample rates and data from audio file
    sample_rate, data = wavfile.read(audio_file)

    global freqs, spectrum, t

    # Create a single subplot with two sets of axes
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 12), sharex=True)

    # Plot the spectrogram
    spectrum, freqs, t, im = ax1.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap("autumn_r"))
    ax1.set_title("Spectrogram")
    ax1.set_ylabel("Frequency (Hz)")

    # Plot reverb time on the second subplot
    data_in_db = frequency_check(frequency)
    ax2.plot(t, data_in_db, linewidth=1, alpha=0.7, color=color)
    ax2.set_title(f"Reverb Time Plot - Frequency: {frequency} Hz")
    ax2.set_xlabel("Time (s)")
    ax2.set_ylabel("Power (dB)")

    # Find an index of a max value
    index_of_max = np.argmax(data_in_db)
    value_of_max = data_in_db[index_of_max]

    # Slice array from a max value
    sliced_array = data_in_db[index_of_max:]

    value_of_max_less_5 = value_of_max - 5
    value_of_max_less_25 = value_of_max - 25

    # Getting index of the closest value in the data to the max value minus 5
    value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)
    index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)

    # Getting index of the closest value in the data to the max value minus 25
    value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)
    index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

    # Plot max, mid, and min points on the reverb time plot
    ax2.plot(t[index_of_max], value_of_max, "go", label="Max")
    ax2.plot(t[index_of_max_less_5], value_of_max_less_5, "yo", label="Max - 5dB")
    ax2.plot(t[index_of_max_less_25], value_of_max_less_25, "ro", label="Max - 25dB")
    ax2.legend()

    # Extraploating rt20 to get rt60
    rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]
    rt60 = 3 * rt20

    debugg(f"The RT60 reverb time for {frequency} Hz is {round(abs(rt60), 2)} seconds")

    plt.show()

def combine_rt60s(frequencies = [], colors = {}):
    for frequency in frequencies:
        target_frequency = find_target_frequency(freqs, frequency)
        index_of_frequency = np.where(freqs == target_frequency) [0][0]
        data_for_frequency = spectrum[index_of_frequency]
        data_in_db = 10 * np.log10(data_for_frequency)

        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color=(colors.get(frequency)), label=f"{frequency} Hz")
        plt.title(f"Reverb Time Plot - Frequency: {frequencies} Hz")
        plt.xlabel("Time (s)")
        plt.ylabel("Power (dB)")
        plt.legend(loc="upper left")

    plt.show()

def calculate_rt60_time(wave_file, frequency):
    # Plot reverb time on the second subplot
    data_in_db = frequency_check(frequency)

    # Find an index of a max value
    index_of_max = np.argmax(data_in_db)
    value_of_max = data_in_db[index_of_max]

    # Slice array from a max value
    sliced_array = data_in_db[index_of_max:]

    value_of_max_less_5 = value_of_max - 5
    value_of_max_less_25 = value_of_max - 25

    # Getting index of the closest value in the data to the max value minus 5
    value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)
    index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)

    # Getting index of the closest value in the data to the max value minus 25
    value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)
    index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

    rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]
    return 3 * rt20

def rt60_difference(wave_file):
    rt60_difference_reduced = int(0)
    frequencies = [250, 1000, 5000]
    for frequency in frequencies:
        rt60_difference_reduced += calculate_rt60_time(wave_file, frequency)

    return (rt60_difference_reduced // 3) - 0.5
